server/conn.go

```diff
@@ -715,6 +785,35 @@ func (cc *clientConn) openSessionAndDoAuth(authData []byte) error {
 	return nil
 }
 
+func (cc *clientConn) checkAuthPlugin(ctx context.Context, authPlugin *string) ([]byte, error) {
+	// Open a context unless this was done before.
+	if cc.ctx == nil {
+		err := cc.openSession()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	userplugin, err := cc.ctx.AuthPluginForUser(&auth.UserIdentity{Username: cc.user, Hostname: cc.peerHost})
```

Comment:
This is far too complex and beyond my comprehension. Could you refactor the code a bit for readability?

Reply:
```
// step1, there is a variable for the auth plugin from the respond;                                                                                                                                                
//        there is a variable for the user plugin from mysql.user table                                                                                                                                            
//        there is a variable for the default auth plugin                                                                                                                                                          
// we can assign a default value for them if they are empty                                                                                                                                                        
                                                                                                                                                                                                                   
userPlugin = AuthPluginForUser()                                                                                                                                                                                   
if userPlugin = "" {                                                                                                                                                                                               
    userPlugin = default                                                                                                                                                                                           
}                                                                                                                                                                                                                  
                                                                                                                                                                                                                   
respPlugin = resp.AuthPlugin                                                                                                                                                                                       
if respPlugin == "" {                                                                                                                                                                                              
    respPlugin = default                                                                                                                                                                                           
}                                                                                                                                                                                                                  
                                                                                                                                                                                                                   
// step2, what we really care is the user plugin should be the same with the response plugin, am I right?                                                                                                              
// They are not empty after step1                                                                                                                                                                                  
                                                                                                                                                                                                                   
if userPlugin != respPlugin {                                                                                                                                                                                      
    resp.Auth = authSwitchRequest(userPlugin)                                                                                                                                                                      
}

// step3, auth data is now resp.Auth, and auth plugin is userPlugin    
```

Reply:
I've updated this, could you check it again?